src/System.Collections/src/System/Collections/Generic/Queue.cs

```diff
@@ -390,22 +386,67 @@ public void Dispose()
             /// <include file='doc\Queue.uex' path='docs/doc[@for="QueueEnumerator.MoveNext"]/*' />
             public bool MoveNext()
             {
-                if (_version != _q._version) throw new InvalidOperationException(SR.InvalidOperation_EnumFailedVersion);
-
-                if (_index == -2)
-                    return false;
-
-                _index++;
-
-                if (_index == _q._size)
+                // Typically both of these conditions will be true, so save ourselves
+                // a branch and use bitwise AND.
+                bool typicalIteration = (_version == _q._version) & (_index != -2);
+
+                // Instead of return true/false, we assign the result to a variable and
+                // return it at the end. This helps decrease code size, as currently the
+                // jit cannot do this for us and generates code for 3 returns.
```

Comment:
This seems like unnecessary complexity.  I'd rather leave it as it was and then see the JIT improve so that this and all other such cases benefit.  This new code is, IMO, much harder to follow than the previous version.